python/src/aiconfig/editor/server/server_utils.py

```diff
@@ -213,3 +220,127 @@ def init_server_state(app: Flask, edit_config: EditServerConfig) -> Result[None,
+
+
+def make_op_run_method(method_name: MethodName) -> Operation[None]:
+    def _op(aiconfig: AIConfigRuntime, operation_args: OpArgs) -> Result[None, str]:
+        LOGGER.info(f"Running method: {method_name}, {operation_args=}")
+        return _safe_run_aiconfig_method(aiconfig, method_name, operation_args)
+
+    return _op
+
+
+def run_aiconfig_operation_with_op_args(
+    aiconfig: AIConfigRuntime | None,
+    operation_name: str,
+    operation: Operation[None],
+    res_op_args: Result[OpArgs, str],
+) -> FlaskResponse:
+    """
+    Use this when you have to do specific logic to map
+    the request JSON to the OpArgs. See "/api/run" for example.
+    Otherwise, use `run_aiconfig_operation_with_request_json`.
+    """
+    if aiconfig is None:
+        LOGGER.info(f"No AIConfig in memory, can't run {operation_name}.")
+        return HttpResponseWithAIConfig(
+            message=f"No AIConfig in memory, can't run {operation_name}.",
+            code=400,
+            aiconfig=None,
+        ).to_flask_format()
+
+    LOGGER.info(f"About to run operation: {operation_name}, {res_op_args=}")
+    res_run = result.do(
+        operation(aiconfig, OpArgs(op_args_ok))
+        #
+        for op_args_ok in res_op_args
+    )
+
+    LOGGER.info(f"{res_run=}")
+
+    match res_run:
+        case Ok(_):
+            return HttpResponseWithAIConfig(
+                message="Done",
+                aiconfig=aiconfig,
+            ).to_flask_format()
+        case Err(e):
+            return HttpResponseWithAIConfig(
+                message=f"Failed run {operation_name}: {e}",
+                code=400,
+                aiconfig=None,
+            ).to_flask_format()
+
+
+def _validated_op_args_from_request_json(
+    request_json: core_utils.JSONObject,
+    signature: dict[str, Type[Any]],
+) -> Result[OpArgs, str]:
+    if signature.keys() != request_json.keys():
+        LOGGER.info(f"Expected keys: {signature.keys()}, got: {request_json.keys()}")
+        return Err(f"Expected keys: {signature.keys()}, got: {request_json.keys()}")
+
+    def _resolve(key: str, value: core_utils.JSONValue, signature: dict[str, Type[Any]]) -> Result[Any, str]:
+        LOGGER.info(f"Resolving: {key}, {value}, {signature[key]}")
+        _type = signature[key]
+        if isinstance(value, typing.Dict):
+            return core_utils.safe_model_validate_json_object(_type, value)
+        else:
+            return Ok(value)
+
+    operation_args_results = {key: _resolve(key, value, signature) for key, value in request_json.items()}
+    LOGGER.info(f"{operation_args_results=}")
+    res_op_args: Result[OpArgs, str] = cast(Result[OpArgs, str], core_utils.result_reduce_dict_all_ok(operation_args_results))
+
+    return res_op_args
+
+
+def run_aiconfig_operation_with_request_json(
+    aiconfig: AIConfigRuntime | None,
+    request_json: core_utils.JSONObject,
+    operation_name: str,
+    operation: Operation[None],
```

Comment:
Instead of passing operation_name string, couldn't this be inferred from the operation (or set the operation name as a static variable in the Operation instance to obtain here)

Reply:
Yes, it could be bundled into the Operation or inferred. I just have it like this for straight-forward logging, but yes it could be refactored. Happy to do that as a follow up https://docs.google.com/document/d/1jUsV3Khu_Gs31dIBxh1n2DnF_AbnRPiN9etZpLe1dk8/edit